golang/pkg/rnr/job.go

```diff
@@ -58,8 +62,81 @@ func (j *Job) Proto(updater func(*pb.Job)) *pb.Job {
 	return ret
 }
 
+// taskDiff recursively walks the task protobufs and calculates any differences
+func taskDiff(path []string, old *pb.Task, new *pb.Task) []string {
+	var ret []string
+
+	oldState := "(new)"
+	newState := "(deleted)"
+	oldMessage := ""
+	newMessage := ""
+
+	if old != nil {
+		oldState = old.State.String()
+		oldMessage = old.GetMessage()
+	}
+	if new != nil {
+		newState = new.State.String()
+		newMessage = new.GetMessage()
+	}
+
+	if oldState != newState || oldMessage != newMessage {
+		ret = append(ret, fmt.Sprintf("[%s]: %s (%s) -> %s (%s)", strings.Join(path, "/"), oldState, oldMessage, newState, newMessage))
+	}
+
+	// Check children
+	childrenMap := make(map[string]bool)
+	oldChildren := make(map[string]*pb.Task)
+	if old != nil {
+		for _, c := range old.Children {
+			childrenMap[c.Name] = true
+			oldChildren[c.Name] = c
+		}
+	}
+
+	newChildren := make(map[string]*pb.Task)
+	if new != nil {
+		for _, c := range new.Children {
+			childrenMap[c.Name] = true
+			newChildren[c.Name] = c
+		}
+	}
+
+	children := []string{}
```

Comment:
A minor optimization but that it is really significant is creating the slice with the right capacity since:
```suggestion
	children := make([]string, 0, len(childrenMap))
```

Reply:
Neat; I didn't realize this :-)